// f5-nodejs/index.js
/*jshint node:true*/
'use strict';

/** @module f5-nodejs */

var ILXServer = require('./lib/ilx_server.js');

var ILXPlugin = require('./lib/ilx_plugin.js');
var ILXPluginOptions = require('./lib/ilx_plugin_options.js');
var ILXLbOptions = require('./lib/ilx_lb_options.js');
var ILXFlow = require('./lib/ilx_flow.js');
var ILXStream = require('./lib/ilx_stream.js');
var ILXTable = require('./lib/ilx_table.js');
var ILXBufferUtil = require('./lib/ilx_buffer_util.js');
var ILXHttpRequest = require('./lib/ilx_http_request.js');

/**
 * @kind class
 * @see {@link ILXServer}
 */
exports.ILXServer = ILXServer;

exports.ILXFlow = ILXFlow;
exports.ILXLbOptions = ILXLbOptions;
exports.ILXPlugin = ILXPlugin;
exports.ILXPluginOptions = ILXPluginOptions;
exports.ILXStream = ILXStream;
exports.ILXTable = ILXTable;
exports.ILXBufferUtil = ILXBufferUtil;
exports.ILXHttpRequest = ILXHttpRequest;

/**
 * @param {object} [methods] - hash of methods in {name: callback} format (optional)
 * @returns {ILXServer} instance of ILXServer
 */
exports.createILXServer = function (methods) {
    return new ILXServer(methods);
};
